File: src/main/java/rars/venus/settings/editor/views/FontSettingsView.java
package rars.venus.settings.editor.views;

import org.jetbrains.annotations.NotNull;
import rars.util.FontUtilities;
import rars.util.FontWeight;
import rars.venus.util.GridBagBuilder;

import javax.swing.*;
import java.awt.*;

public final class FontSettingsView extends JPanel {
    private static final @NotNull JLabel
        FONT_LABEL = new JLabel("Font family"),
        WEIGHT_LABEL = new JLabel("Font weight"),
        SIZE_LABEL = new JLabel("Font size"),
        LIGATURES_LABEL = new JLabel("Ligatures");

    public final @NotNull JSpinner fontSizeSpinner;
    public final @NotNull JComboBox<@NotNull String> fontSelector;
    public final @NotNull JComboBox<@NotNull FontWeight> fontWeightSelector;
    public final @NotNull JCheckBox ligaturesCheckbox;

    public FontSettingsView() {
        super(new GridBagLayout());
        final var builder = new GridBagBuilder(this)
            .withInsets(new Insets(5, 5, 5, 5));

        // font family
        final var fontFamilies = FontUtilities.allFontFamilies;
        this.fontSelector = new JComboBox<>(fontFamilies);
        this.fontSelector.setEditable(false);
        this.fontSelector.setMaximumRowCount(20);
        builder.addLabelAndField(0, FONT_LABEL, this.fontSelector);

        // font size
        final var fontSizeModel = new SpinnerNumberModel(
            12,
            1,
            100,
            1
        );
        this.fontSizeSpinner = new JSpinner(fontSizeModel);
        this.fontSizeSpinner.setToolTipText("Current font size in points.");
        builder.addLabelAndField(1, SIZE_LABEL, this.fontSizeSpinner);

        // font weight
        this.fontWeightSelector = new JComboBox<>(FontWeight.values());
        this.fontWeightSelector.setEditable(false);
        this.fontWeightSelector.setMaximumRowCount(FontWeight.values().length);
        builder.addLabelAndField(2, WEIGHT_LABEL, this.fontWeightSelector);

        // ligatures
        this.ligaturesCheckbox = new JCheckBox();
        this.ligaturesCheckbox.setToolTipText("Enable or disable ligatures.");
        builder.addLabelAndField(3, LIGATURES_LABEL, this.ligaturesCheckbox);

        // filler to push content to top
        builder.addFiller(4, 2);
    }
}

File: src/main/java/rars/venus/settings/editor/views/FontSettingsView.kt
package rars.venus.settings.editor.views

import rars.util.FontUtilities
import rars.util.FontWeight
import rars.venus.util.GridBagLayout
import java.awt.GridBagConstraints
import java.awt.Insets
import javax.swing.*

class FontSettingsView : JPanel() {
    val fontSizeSpinner = JSpinner().apply {
        model = SpinnerNumberModel(
            /* value = */ 12,
            /* minimum = */ 1,
            /* maximum = */ 100,
            /* stepSize = */ 1
        )
        toolTipText = "Current font size in points."
    }
    val fontSelector: JComboBox<String> =
        JComboBox(FontUtilities.allFontFamilies).apply {
            isEditable = false
            maximumRowCount = 20
        }
    val fontWeightSelector: JComboBox<FontWeight> =
        JComboBox(FontWeight.entries.toTypedArray()).apply {
            isEditable = false
            maximumRowCount = FontWeight.entries.size
        }
    val ligaturesCheckbox = JCheckBox().apply {
        toolTipText = "Enable or disable ligatures."
    }

    init {
        val baseConstraints = GridBagConstraints().apply {
            fill = GridBagConstraints.BOTH
            insets = Insets(5, 5, 5, 5)
        }
        GridBagLayout(baseConstraints) {
            add(FONT_LABEL, gridx = 0, gridy = 0)
            add(fontSelector, gridx = 1, gridy = 0)

            add(SIZE_LABEL, gridx = 0, gridy = 1)
            add(fontSizeSpinner, gridx = 1, gridy = 1)

            add(WEIGHT_LABEL, gridx = 0, gridy = 2)
            add(fontWeightSelector, gridx = 1, gridy = 2)

            add(LIGATURES_LABEL, gridx = 0, gridy = 3)
            add(ligaturesCheckbox, gridx = 1, gridy = 3)
        }
    }

    companion object {
        private val FONT_LABEL = JLabel("Font family")
        private val WEIGHT_LABEL = JLabel("Font weight")
        private val SIZE_LABEL = JLabel("Font size")
        private val LIGATURES_LABEL = JLabel("Ligatures")
    }
}